pkg/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,394 @@
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	allocation, exists := c.ipAllocationMap[service]
+	if exists {
+		c.externalIPAllocator.ReleaseIP(allocation.ipPool, allocation.ip)
+		delete(c.ipAllocationMap, service)
+	}
+}
+
+func (c *ServiceExternalIPController) getAssignedExternalIPAllocation(service apimachinerytypes.NamespacedName) (*ipAllocation, bool) {
+	c.ipAllocationMutex.RLock()
+	defer c.ipAllocationMutex.RUnlock()
+	allocation, exist := c.ipAllocationMap[service]
+	return allocation, exist
+}
+
+func getServiceExternalIP(service *corev1.Service) string {
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		return ""
+	}
+	return service.Status.LoadBalancer.Ingress[0].IP
+}
+
+func (c *ServiceExternalIPController) syncService(key apimachinerytypes.NamespacedName) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Service for %s. (%v)", key, time.Since(startTime))
+	}()
+
+	prev, err := c.serviceLister.Services(key.Namespace).Get(key.Name)
+	if err != nil {
+		// Service already deleted
+		if apimachineryerrors.IsNotFound(err) {
+			c.deleteExternalIP(key)
+			return nil
+		}
+		return err
+	}
+
+	service := prev.DeepCopy()
+	// Service does not need external IP or type has changed.
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		c.deleteExternalIP(key)
+		return nil
+	}
+
+	currentIPPool := service.ObjectMeta.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey]
+	prevIPAllocation, allocationExists := c.getAssignedExternalIPAllocation(key)
+	currentExternalIP := getServiceExternalIP(service)
+
+	// If user specifies external IP in spec, we should check whether it matches the current external IP.
+	specIPMatched := service.Spec.LoadBalancerIP == "" || service.Spec.LoadBalancerIP == currentExternalIP
+
+	if allocationExists && specIPMatched &&
+		c.externalIPAllocator.IPPoolExists(currentIPPool) &&
+		c.externalIPAllocator.IPPoolHasIP(currentIPPool, prevIPAllocation.ip) &&
+		currentIPPool == prevIPAllocation.ipPool &&
+		currentExternalIP == prevIPAllocation.ip.String() {
+		return nil
+	}
+
+	// The ExternalIPPool does not exist or has been deleted. Reclaim the external IP.
+	if currentIPPool != "" && !c.externalIPAllocator.IPPoolExists(currentIPPool) {
+		c.deleteExternalIP(key)
+		if currentExternalIP != "" {
+			service.Status.LoadBalancer.Ingress = nil
+			return c.updateService(prev, service)
+		}
+	}
+
+	// the external IP or ExternalIPPool changed somehow. Delete the previous allocation.
+	c.deleteExternalIP(key)
+	var newIPPool string
+	var newExternalIP net.IP
+	if service.Spec.LoadBalancerIP != "" {
+		// find the coressponding ExternalIPPool for user specified IP address.
+		newExternalIP = net.ParseIP(service.Spec.LoadBalancerIP)
+		newIPPool, err = c.externalIPAllocator.LocateIP(newExternalIP)
+		if err == nil {
+			err = c.externalIPAllocator.UpdateIPAllocation(newIPPool, newExternalIP)
```

Comment:
Could the two calls be merged to one call `AllocateIP(ip net.IP)`? If ip is not specified, it allocates next available one. This may save code on both caller and callee sides

Reply:
updated per suggestion.